lib/src/client/session/services/subscriptions/mod.rs

```diff
@@ -0,0 +1,451 @@
+                discard_oldest: i.discard_oldest,
+                filter: i.filter,
+            };
+
+            let client_handle = monitored_item.client_handle();
+            let monitored_item_id = monitored_item.id();
+            self.monitored_items
+                .insert(monitored_item_id, monitored_item);
+            self.client_handles.insert(client_handle, monitored_item_id);
+        });
+    }
+
+    pub(crate) fn modify_monitored_items(&mut self, items_to_modify: &[ModifyMonitoredItem]) {
+        items_to_modify.iter().for_each(|i| {
+            if let Some(ref mut monitored_item) = self.monitored_items.get_mut(&i.id) {
+                monitored_item.set_sampling_interval(i.sampling_interval);
+                monitored_item.set_queue_size(i.queue_size as usize);
+            }
+        });
+    }
+
+    pub(crate) fn delete_monitored_items(&mut self, items_to_delete: &[u32]) {
+        items_to_delete.iter().for_each(|id| {
+            // Remove the monitored item and the client handle / id entry
+            if let Some(monitored_item) = self.monitored_items.remove(id) {
+                let _ = self.client_handles.remove(&monitored_item.client_handle());
+            }
+        })
+    }
+
+    pub(crate) fn set_triggering(
+        &mut self,
+        triggering_item_id: u32,
+        links_to_add: &[u32],
+        links_to_remove: &[u32],
+    ) {
+        if let Some(ref mut monitored_item) = self.monitored_items.get_mut(&triggering_item_id) {
+            monitored_item.set_triggering(links_to_add, links_to_remove);
+        }
+    }
+
+    pub(crate) fn on_notification(
+        &mut self,
+        notification: NotificationMessage,
+        decoding_options: &DecodingOptions,
+    ) {
+        let Some(notifications) = notification.notification_data else {
+            return;
+        };
+
+        for obj in notifications {
+            if obj.node_id.namespace != 0 {
+                continue;
+            }
+
+            let Identifier::Numeric(id) = obj.node_id.identifier else {
+                continue;
+            };
+
+            if id == ObjectId::DataChangeNotification_Encoding_DefaultBinary as u32 {
+                match obj.decode_inner::<DataChangeNotification>(decoding_options) {
+                    Ok(it) => {
+                        for notif in it.monitored_items.into_iter().flatten() {
+                            let item = self
+                                .client_handles
+                                .get(&notif.client_handle)
+                                .and_then(|handle| self.monitored_items.get(handle));
+
+                            if let Some(item) = item {
+                                self.callback.on_data_value(notif.value, &item);
+                            }
+                        }
+                    }
+                    Err(e) => warn!("Failed to decode data change notification: {e}"),
+                }
```

Comment:
Can I ask the reasoning behind changing this logic as compared to how the callback worked in the old client?

While I understand that for some use cases it might be (slightly) easier to get each value on its own (as compared to having to loop through them), is does however remove some context and prevents being able to batch process values received from a single publish response.

In our case this actually breaks some of our logic and instead of being able to use and rely on the values returned by the publish responses, we now have to use an arbitrary de-bounce delay to group/batch some received values. But especially with low publish intervals this now almost becomes a continuous stream instead of batched responses.

As I mentioned I understand that depending on how you want to process or store the values the new approach might be slightly handier, but it prevents the "batched" approaches. While if we just forward the batch that we received we would allow people to make that choice on their own and either process the vales as a batch or loop over them and process them one by one.

So long story short, unless there is a very compelling reason for having it this way I would like to make a PR to change this behavior to how it worked before. Thanks!

Reply:
Yeah, sure, if you want to change it back that is probably fine. If you do that, mind implementing the current approach as a layer on top of the more simple API you suggest?

For a lot of simple applications it's convenient to have an easy to use and, importantly, easy to understand interface for consuming notifications.

Reply:
The "easy to understand" part is the main justification behind the change. Consuming subscriptions is _the_ thing you will want to do with a starter application, and I would like it to be as simple as possible.

That said, it should be easy enough to just have to traits, one for your low level callbacks, and one for my high level, then blanket implement the former for the latter.

Reply:
Yeah fully agree... As said I can see use cases for both approaches so sounds like a good idea to just offer them both instead of just one. Will have a look at it somewhere in the coming week 👍🏻